File: Algorithm_103_Binary Tree Zigzag Level Order Traversal.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
    struct node {
        int level;
        TreeNode *n;
        node(TreeNode *nod, int l) : level(l), n(nod) {}
    };
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int> > result;
        if (!root) return result;
        queue<node> q;
        node head(root, 0);
        int count = 0;
        q.push(head);
        vector<int> tmpRes;
        while (!q.empty())
        {
            node top = q.front();
            q.pop();
            if (top.level != count)
            {
                
                count = top.level;
                if (count % 2 == 0) 
                {
                    vector<int> reversed;
                    for (int i = tmpRes.size() - 1; i >= 0; i--) reversed.push_back(tmpRes[i]);
                    result.push_back(reversed);
                    reversed.clear();
                }
                else result.push_back(tmpRes);
                tmpRes.clear();
            }
            tmpRes.push_back(top.n->val);
            if (top.n->left)
            {
                node tmp(top.n->left, top.level + 1);
                q.push(tmp);
            } 
            if (top.n->right)
            {
                node tmp(top.n->right, top.level + 1);
                q.push(tmp);
            }
        }
        if (tmpRes.size() && count % 2 == 0) result.push_back(tmpRes);
        else 
        {
            vector<int> reversed;
            for (int i = tmpRes.size() - 1; i >= 0; i--) reversed.push_back(tmpRes[i]);
            result.push_back(reversed);
        }
        return result;
    }
};

File: Algorithm_103_Binary Tree Zigzag Level Order Traversal.java
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// My Solution
// 思路：用变量flag记录当前应该正序还是逆序，每次遍历完当前层，即将进入遍历下一层时，就把flag反转。flag为true表示正序遍历，为false表示逆序遍历
// reverse方法负责把List中的元素反转

class Solution {

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        boolean flag = true;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> curLayer = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode curNode = queue.poll();
                if (curNode == null) continue;
                curLayer.add(curNode.val);
                queue.offer(curNode.left);
                queue.offer(curNode.right);
            }
            if (!flag) reverse(curLayer);
            if (curLayer.size() > 0) res.add(curLayer);
            flag = !flag;
        }
        return res;
    }

    private void reverse(List<Integer> arr) {
        for (int i = 0; i < arr.size() / 2; i++) {
            int tmp = arr.get(i);
            arr.set(i, arr.get(arr.size() - i - 1));
            arr.set(arr.size() - i - 1, tmp);
        }
    }

}

// LinkedList Solution
// 思路：利用链表的在头部插入时间复杂度为O(1)的性质

class LinkedListSolution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        boolean flag = true;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> curLayer = new LinkedList<>(); // use LinkedList instead of ArrayList
            for (int i = 0; i < size; i++) {
                TreeNode curNode = queue.poll();
                if (curNode == null) continue;
                if (flag) curLayer.add(curNode.val);
                else curLayer.add(0, curNode.val); // insert at the front of the list
                queue.offer(curNode.left);
                queue.offer(curNode.right);
            }
            if (curLayer.size() > 0) res.add(curLayer);
            flag = !flag;
        }
        return res;
    }
}